driver/utils/type_info.h

```diff
@@ -0,0 +1,2112 @@
+            constexpr std::uint32_t dec_mult = 10;
+            constexpr std::uint32_t byte_mult = 1 << 8;
+
+            dest.sign = src.sign;
+
+            if (dest.precision == 0) {
+                dest.precision = src.precision;
+                dest.scale = src.scale;
+            }
+
+            auto tmp_src = src;
+
+            // Adjust the detected scale if needed.
+
+            while (tmp_src.scale < dest.scale) {
+                if ((src_value_max / dec_mult) < tmp_src.value)
+                    throw std::runtime_error("Cannot fit source Numeric value into destination Numeric specification: value is too big for internal representation");
+
+                tmp_src.value *= dec_mult;
+                ++tmp_src.scale;
+            }
+
+            while (dest.scale < tmp_src.scale) {
+                tmp_src.value /= dec_mult;
+                --tmp_src.scale;
+            }
+
+            // Transfer the value.
+
+            for (std::size_t i = 0; tmp_src.value != 0; ++i) {
+                if (i >= lengthof(dest.val) || i > dest.precision)
+                    throw std::runtime_error("Cannot fit source Numeric value into destination Numeric specification: value is too big for ODBC Numeric representation");
+
+                dest.val[i] = static_cast<std::uint32_t>(tmp_src.value % byte_mult);
+                tmp_src.value /= byte_mult;
+            }
+        }
+    };
+
+    template <>
+    struct from_value<DataSourceType<DataSourceTypeId::Decimal32>> {
+        using SourceType = DataSourceType<DataSourceTypeId::Decimal32>;
+
+        template <typename DestinationType>
+        struct to_value {
+            static inline void convert(const SourceType & src, DestinationType & dest) {
+                return from_value<DataSourceType<DataSourceTypeId::Decimal>>::template to_value<DestinationType>::convert(src, dest);
+            }
+        };
+    };
+
+    template <>
+    struct from_value<DataSourceType<DataSourceTypeId::Decimal64>> {
+        using SourceType = DataSourceType<DataSourceTypeId::Decimal64>;
+
+        template <typename DestinationType>
+        struct to_value {
+            static inline void convert(const SourceType & src, DestinationType & dest) {
+                return from_value<DataSourceType<DataSourceTypeId::Decimal>>::template to_value<DestinationType>::convert(src, dest);
+            }
+        };
+    };
+
+    template <>
+    struct from_value<DataSourceType<DataSourceTypeId::Decimal128>> {
+        using SourceType = DataSourceType<DataSourceTypeId::Decimal128>;
+
+        template <typename DestinationType>
+        struct to_value {
+            static inline void convert(const SourceType & src, DestinationType & dest) {
+                return from_value<DataSourceType<DataSourceTypeId::Decimal>>::template to_value<DestinationType>::convert(src, dest);
+            }
+        };
+    };
+
+    template <typename DestinationType>
+    struct to_buffer {
```

Comment:
Same thing applies to `to_buffer<X>::from_value<Y>::convert`, maybe `convertFromValueToBuffer(const Y & y, X & x)` ?

Reply:
Even more, `to_buffer` breaks pattern:

`from_Something<X>::to_SOMETHING_ELSE<Y>::convert(x, y)`
to opposite:
`to_SOMETHING<X>::from_SOMETHING_ELSE<X>::convert(y, x)`

and that complicates things even further.

Reply:
That's severely affect maintainability, as mentioned earlier.

Reply:
How does it brake the pattern?
```
to_value<OfType>
from_value<OfType>
to_buffer<OfType>
from_buffer<OfType>
```

Reply:
ok, so `to_smth` and `from_smt` defined and used differently, according to the way the  conversion are used, and that allows to define some default conversions that cover more cases.

Reply:
By interrupting the flow, most of the code here converters from **left** type to **right** type, and `to_buffer` is an exception. What I am trying to say: this is already complicated enough, no need to make it even harder to grasp.

Reply:
That's all about compromises. Preserving `to_*` -> `from_*` structure may result in more actual code, that does pretty much the same thing.